plugin-api/src/main/java/org/sonar/api/security/ExternalGroupsProvider.java

```diff
@@ -45,19 +46,46 @@ public Collection<String> doGetGroups(Context context) {
 
   public static final class Context {
     private String username;
+    /**
+     * @deprecated since 9.16 use {@link #httpRequest} instead
+     */
+    @Deprecated(since = "9.16", forRemoval = true)
     private HttpServletRequest request;
+    private HttpRequest httpRequest;
 
+    /**
+     * @deprecated since 9.16 use {@link #Context(String, HttpRequest)} instead
+     */
+    @Deprecated(since = "9.16", forRemoval = true)
     public Context(String username, HttpServletRequest request) {
       this.username = username;
       this.request = request;
     }
 
+    public Context(String username, HttpRequest httpRequest) {
+      this.username = username;
+      this.httpRequest = httpRequest;
+    }
+
     public String getUsername() {
       return username;
     }
 
+    /**
+     * @deprecated since 9.16
+     * Use this only if context was created with {@link #Context(String, HttpServletRequest)}
```

Comment:
Consumers of the context shouldn't need to know how it was instantiated by the product implementing the API.
I suggest we have both old and new implementations available during the deprecation period.

Same applies to the other Context.

Reply:
Actually, I may have misunderstood your comment. I pushed a fixup; please let me know if it makes sense.

Reply:
@dbmeneses we will merge for now as we think it is fine now. Let us know if you think we should change something here.